### ISSUE TITLE
Pretrained models have no node names

### ISSUE THREAD

**Issue by markorakita** [Microsoft/CNTK#2585]
I've tried to use pretrained models from here to extract features from some of intermediate layers: https://github.com/Microsoft/CNTK/blob/master/PretrainedModels

But the problem is that none of those models are trained with properly named layers. When I try to do Function.FindByName(layerName) in C# managed API, I can't find any layer.

Is there any other way to extract Function for certain intermediate layer instead of by name?

**Comment by aWeinzierl** [comment]
If I am not wrong, you can just use model.Arguments[x] to access any layer of your model, after you loaded it.
If you want to programmatically find a specific one you still could use something like:
var wantedLayer = model.Arguments.Single(layer => layer.dimensions==previouslyDefinedDimensions);
to extract a specific layer by different properties.

**Comment by aWeinzierl** [comment]
sorry, my suggestion will not work. I falsely assumed that the functions would be listed in Arguments.

So that is a really good question.

**Comment by haixpham** [comment]
In python API, I use 
```python
nodes = Function.find_all_with_name("")
```
then print all nodes to identify layers

**Comment by aWeinzierl** [comment]
You can do that in C#, too, but it does not include layers which have a name you do not know (e.g. "f.g.ss.f"), does it?

**Comment by haixpham** [comment]
Yeah, thats's true, if the node has a name then it's not included in the list.

@cntk deveopers: it would be really convenient to analyze a network if cntk.logging.graph.plot prints out node names together with the computational graph.

**Comment by YanYas** [comment]
Hi @CNTK devs. This is still a pretty major issue when trying to experiment with layers of a graph. Bump.

**Comment by tjrileywisc** [comment]
I was hoping to use this to leverage pretrained keras models but the .combine() method (on the loaded keras model, when using the CNTK backend) seems to purge all layer names except for the input layer. Faster R-CNN wants you to extract portions of the network to specify last convolution layer blocks, I don't see how to do this without having control over the layer names.

**Comment by szabolcsvelkei** [comment]
This is a terrible ugly solution, but you can iterate through the nodes and write down their names:    

    `
            foreach (var node in Model.Outputs)
            {
                WriteAllNodes(node);
            }
    `

 And:

    `
        private void WriteAllNodes(Variable node)
        {
            foreach (var inputNode in node.Owner.Inputs)
                if (inputNode.Kind == VariableKind.Output)
                {
                    Console.WriteLine(inputNode.Name);
                    WriteAllNodes(inputNode);
                    break;
                }
        }
    `

**Comment by nimning** [comment]
Same issue here. Not name for each layer in the pretrained model. Make transfer learning very difficult.
